[Bug] Error CS0234 when Upgrading from 2.61.0 to 2.62.xx

**Describe the bug**

I wanted to migrate the Docfx Version from 2.59.2 to 2.62.1 but apperently in addition to issue #8428, I can not build documentation from source code (.cs file extensions). The Error CS0234 appears, which means that he does not find the namespace. Version 2.61.0 works as exptected and I can build the automated API documentation via .cs file extensions with the same `docfx.json` which I used for 2.59.2. When I build the API documentation with the .sln file or the .csproj extension in Version 2.62.xx, than it works too, but I do not need complete documentation for all namespaces and classes. I can use Filter APIs to exclude the unnecessary namespaces if I work with the .sln, but that does not seem right if it worked in version 2.61.0. For myself, I can not say what change from the release notes in version 2.62.0 broke the expected behavior.

**To Reproduce**
Steps to reproduce the behavior:
1. Project with API documentation for specific .cs files
2. docfx
3. docfx serve _site

**Expected behavior**
Functional output like in version 2.61.0

<img width="1215" alt="Bildschirm­foto 2023-02-21 um 11 48 39" src="https://user-images.githubusercontent.com/78415305/220328754-4f2a3cee-48d6-4503-bdc7-ec2ffdef24ba.png">

**Context (please complete the following information):**
- OS: Windows
- Docfx version: [e.g. 2.62.x]

**Additional context**
Unfortunately, I can not send a .zip file with the complete project and the two solutions which I try to build an automated API for because of my work regulations but you can see in the following the `docfx.json` which works in version 2.59.2 and 2.61.0 but is broken in 2.62.x. In the picture, you see the msbuild error CS0234.

### Error CS0234

<img width="1023" alt="Bildschirm­foto 2023-02-21 um 11 51 04" src="https://user-images.githubusercontent.com/78415305/220328830-2285a2d3-9dad-4d50-868b-9b52c6a1bfb9.png">

### docfx..json
```
{
  "metadata": [{
      "src": [{
        "files": ["src/AdamWS/**/ADAMAPIv01.asmx.cs", "src/AdamWS/**/IdmAPIv01.asmx.cs", "src/AdamWS/**/DataObjects/**.cs"],
        "exclude": ["_site/**", "**/bin/**", "**/obj/**", "**/[Tt]ests/**"]
      }],
      "dest": "adam/soap-api",
      "disableGitFeatures": false,
      "disableDefaultFilter": false
    },
    {
      "src": [{
        "files": ["src/CachedExchangeWS/**/AGAPIv02.asmx.cs", "src/CachedExchangeWS/**/AGAPIv02_1.asmx.cs", "src/CachedExchangeWS/**/APIv05_1.asmx.cs", "src/CachedExchangeWS/**/APIv05_2.asmx.cs", "src/CachedExchangeWS/**/DataObjects/*.cs"],
        "exclude": ["_site/**", "**/bin/**", "**/obj/**", "**/[Tt]ests/**"]
      }],
      "dest": "mail/soap-api",
      "disableGitFeatures": false,
      "disableDefaultFilter": false
    }
  ],
  "build": {
    "content": [{
        "files": ["adam/**.yml", "adam/**.md"]
      },
      {
        "files": ["mail/**.md", "mail/**.yml"]
      },
	  {
        "files": ["toc.yml", "index.md"]
      },
      {
        "files": ["changelog/**.md", "changelog/**/toc.yml"]
      },
      {
        "exclude": ["_site/**", "**/bin/**", "**/obj/**", "**/[Tt]ests/**"]
      }
    ],
    "resource": [{
      "files": ["images/**"]
    }],
    "overwrite": [{
      "files": ["_site/**.md"],
      "exclude": ["obj/**", "_site/**"]
    }],
    "dest": "_site",
    "globalMetadataFiles": [],
    "fileMetadataFiles": [],
    "template": ["default", "templates/singulinkfx", "plugins/memberpage.2.59.0/content"],
    "postProcessors": [],
    "markdownEngineName": "markdig",
    "noLangKeyword": false,
    "keepFileLink": false,
    "cleanupCacheHistory": false,
    "disableGitFeatures": false,
    "globalMetadata": {
      "_appTitle": "IT Center RWTH Aachen University",
      "_appName": "Doc Portal",
      "_appFaviconPath": "images/favicon.ico",
      "_appLogoPath": "images/rwth-itc.png",
      "_appFooter": "<strong>© IT Center RWTH Aachen University. All rights reserved.</strong>",
      "_copyrightFooter": " ",
      "_enableSearch": true,
      "_disableSideFilter": false,
      "_enableNewTab": true,
      "_disableContribution": false,
      "_disableBreadcrumb": false
    }
  }
}
```

--- a/src/Microsoft.DocAsCode.Dotnet/CompilationHelper.cs
+++ b/src/Microsoft.DocAsCode.Dotnet/CompilationHelper.cs
@@ -25,7 +25,7 @@ internal static class CompilationHelper
                 """),
     };
 
-    public static bool CheckDiagnostics(this Compilation compilation)
+    public static bool CheckDiagnostics(this Compilation compilation, bool errorAsWarning)
     {
         var errorCount = 0;
 
@@ -34,7 +34,8 @@ internal static class CompilationHelper
             if (diagnostic.IsSuppressed)
                 continue;
 
-            if (diagnostic.Severity is DiagnosticSeverity.Warning)
+            if (diagnostic.Severity is DiagnosticSeverity.Warning ||
+                (diagnostic.Severity is DiagnosticSeverity.Error && errorAsWarning))
             {
                 Logger.LogWarning(diagnostic.ToString());
                 continue;

--- a/src/Microsoft.DocAsCode.Dotnet/DotnetApiCatalog.cs
+++ b/src/Microsoft.DocAsCode.Dotnet/DotnetApiCatalog.cs
@@ -126,6 +126,7 @@ public static class DotnetApiCatalog
             DisableDefaultFilter = configModel?.DisableDefaultFilter ?? false,
             NoRestore = configModel?.NoRestore ?? false,
             NamespaceLayout = configModel?.NamespaceLayout ?? NamespaceLayout.Flattened,
+            AllowCompilationErrors = configModel?.AllowCompilationErrors ?? false,
             Files = expandedFiles.Items.SelectMany(s => s.Files).ToList(),
             References = expandedReferences?.Items.SelectMany(s => s.Files).ToList(),
         };

--- a/src/Microsoft.DocAsCode.Dotnet/ExtractMetadata/ExtractMetadataConfig.cs
+++ b/src/Microsoft.DocAsCode.Dotnet/ExtractMetadata/ExtractMetadataConfig.cs
@@ -32,4 +32,6 @@ internal class ExtractMetadataConfig
     public NamespaceLayout NamespaceLayout { get; init; }
 
     public Dictionary<string, string> MSBuildProperties { get; init; }
+
+    public bool AllowCompilationErrors { get; init; }
 }

--- a/src/Microsoft.DocAsCode.Dotnet/ExtractMetadata/ExtractMetadataWorker.cs
+++ b/src/Microsoft.DocAsCode.Dotnet/ExtractMetadata/ExtractMetadataWorker.cs
@@ -101,21 +101,21 @@ internal class ExtractMetadataWorker : IDisposable
 
         foreach (var compilation in projectCompilations)
         {
-            hasCompilationError |= compilation.CheckDiagnostics();
+            hasCompilationError |= compilation.CheckDiagnostics(_config.AllowCompilationErrors);
             assemblies.Add((compilation.Assembly, compilation));
         }
 
         if (_files.TryGetValue(FileType.CSSourceCode, out var csFiles))
         {
             var compilation = CompilationHelper.CreateCompilationFromCSharpFiles(csFiles.Select(f => f.NormalizedPath));
-            hasCompilationError |= compilation.CheckDiagnostics();
+            hasCompilationError |= compilation.CheckDiagnostics(_config.AllowCompilationErrors);
             assemblies.Add((compilation.Assembly, compilation));
         }
 
         if (_files.TryGetValue(FileType.VBSourceCode, out var vbFiles))
         {
             var compilation = CompilationHelper.CreateCompilationFromVBFiles(vbFiles.Select(f => f.NormalizedPath));
-            hasCompilationError |= compilation.CheckDiagnostics();
+            hasCompilationError |= compilation.CheckDiagnostics(_config.AllowCompilationErrors);
             assemblies.Add((compilation.Assembly, compilation));
         }
 
@@ -125,7 +125,7 @@ internal class ExtractMetadataWorker : IDisposable
             {
                 Logger.LogInfo($"Loading assembly {assemblyFile.NormalizedPath}");
                 var (compilation, assembly) = CompilationHelper.CreateCompilationFromAssembly(assemblyFile.NormalizedPath, _config.References);
-                hasCompilationError |= compilation.CheckDiagnostics();
+                hasCompilationError |= compilation.CheckDiagnostics(_config.AllowCompilationErrors);
                 assemblies.Add((assembly, compilation));
             }
         }

--- a/src/Microsoft.DocAsCode.Dotnet/MetadataJsonItemConfig.cs
+++ b/src/Microsoft.DocAsCode.Dotnet/MetadataJsonItemConfig.cs
@@ -57,4 +57,7 @@ internal class MetadataJsonItemConfig
 
     [JsonProperty("namespaceLayout")]
     public NamespaceLayout NamespaceLayout{  get; set; }
+
+    [JsonProperty("allowCompilationErrors")]
+    public bool AllowCompilationErrors { get; set; }
 }

--- a/test/Microsoft.DocAsCode.Dotnet.Tests/GenerateMetadataFromCSUnitTest.cs
+++ b/test/Microsoft.DocAsCode.Dotnet.Tests/GenerateMetadataFromCSUnitTest.cs
@@ -3656,19 +3656,19 @@ namespace Test1
     {
         var code =
             """
-                namespace Test
+            namespace Test
+            {
+                internal class Foo : IFoo
                 {
-                    internal class Foo : IFoo
-                    {
-                        internal void M1();
-                        protected internal void M2();
-                        private protected void M3();
-                        private void M4();
-                    }
-
-                    internal interface IFoo { }
+                    internal void M1();
+                    protected internal void M2();
+                    private protected void M3();
+                    private void M4();
                 }
-                """;
+
+                internal interface IFoo { }
+            }
+            """;
 
         var output = Verify(code, new() { IncludePrivateMembers = true });
         var foo = output.Items[0].Items[0];
@@ -3680,5 +3680,21 @@ namespace Test1
             "private protected void M3()",
             "private void M4()",
         }, foo.Items.Select(item => item.Syntax.Content[SyntaxLanguage.CSharp]));
+    }
+
+    [Fact]
+    public void TestAllowCompilationErrors()
+    {
+        var code =
+            """
+            namespace Test
+            {
+                public class Foo : Bar {}
+            }
+            """;
+
+        var output = Verify(code, new() { AllowCompilationErrors = true });
+        var foo = output.Items[0].Items[0];
+        Assert.Equal("public class Foo : Bar", foo.Syntax.Content[SyntaxLanguage.CSharp]);
     }
 }

